src/portal/lib/src/config/system/system-settings.component.ts

```diff
@@ -234,6 +291,73 @@ export class SystemSettingsComponent implements OnChanges, OnInit {
         }
         this.config.robot_token_duration.value = +v * ONE_DAY_MINUTES;
     }
-
-
+    deleteItem(index: number) {
+        this.systemWhitelist.items.splice(index, 1);
+    }
+    addToSystemWhitelist() {
+        // remove duplication and add to systemWhitelist
+        let map = {};
+        this.systemWhitelist.items.forEach(item => {
+            map[item.cve_id] = true;
+        });
+        this.cveIds.split(/[\n,]+/).forEach(id => {
+            let cveObj: any = {};
+            cveObj.cve_id = id.trim();
+            if (!map[cveObj.cve_id]) {
+                map[cveObj.cve_id] = true;
+                this.systemWhitelist.items.push(cveObj);
+            }
+        });
+        // clear modal and close modal
+        this.cveIds = null;
+        this.showAddModal = false;
+    }
+    get whitelistHasChange(): boolean {
+        return !compareValue(this.systemWhitelistOrigin, this.systemWhitelist);
+    }
+    canAdd(): boolean {
+        if (this.cveIds) {
+            let arr = this.cveIds.split(/[\n,]+/);
+            let flag = true;
+            for (let i = 0; i < arr.length; i++) {
+                let id = arr[i].trim();
+                if (!/^CVE-[\d]+-[\d]+$/.test(id)) {
+                    flag = false;
+                    break;
+                }
+            }
+            return flag;
+        }
+        return false;
+    }
+    get expiresDate() {
+        if (this.systemWhitelist && this.systemWhitelist.expires_at) {
+            return new Date(this.systemWhitelist.expires_at * 1000);
+        }
+        return null;
+    }
+    set expiresDate(date) {
+        if (this.systemWhitelist && date) {
+            this.systemWhitelist.expires_at = Math.floor(date.getTime() / 1000);
+        }
+    }
+    get neverExpires(): boolean {
+        return !(this.systemWhitelist && this.systemWhitelist.expires_at);
+    }
+    set neverExpires(flag) {
+        if (flag) {
+            this.systemWhitelist.expires_at = null;
+            if (this.dateInput) {
+                this.dateInput.nativeElement.value = null;
+            }
+        } else {
```

Comment:
This can be abstract to s method. Set the Data as a parameter, I see it is used in 2 places.